tests/samplers/test_tfdata_sampler.py

```diff
@@ -0,0 +1,180 @@
+from collections import defaultdict
+
+import tensorflow as tf
+
+from tensorflow_similarity.samplers import tfdata_sampler as tfds
+
+
+class TestCreateGroupedDataset(tf.test.TestCase):
+    def setUp(self):
+        self.ds = tf.data.Dataset.from_tensor_slices(
+            (
+                tf.constant([1, 2, 3, 4, 5, 6, 7, 8], dtype=tf.float32),
+                tf.constant([1, 1, 2, 2, 3, 3, 4, 4], dtype=tf.int64),
+            )
+        )
+
+    def test_returns_correct_number_of_datasets(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds)
+        self.assertLen(cid_datasets, 4)
+
+    def test_returns_correct_number_of_datasets_with_class_list(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds, class_list=[1, 2])
+        self.assertLen(cid_datasets, 2)
+
+    def test_returns_correct_number_of_datasets_with_total_examples(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds, total_examples=1)
+        self.assertLen(cid_datasets, 4)
+
+        # test that each cid dataset has only 1 example that is repeated.
+        for elem in cid_datasets:
+            self.assertAllEqual(
+                list(elem.take(1).as_numpy_iterator()),
+                list(elem.take(1).as_numpy_iterator()),
+            )
+
+    def test_returns_correct_number_of_datasets_with_buffer_size(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds, buffer_size=2)
+        self.assertEqual(len(cid_datasets), 4)
+
+    def test_datasets_repeat(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds)
+        for cid_ds in cid_datasets:
+            self.assertTrue(cid_ds.element_spec, self.ds.element_spec)
+
+            # Check that repeating groups of 2 elements all equal each other.
+            elements = list(cid_ds.take(6).as_numpy_iterator())
+            self.assertAllEqual(elements[:2], elements[2:4])
+            self.assertAllEqual(elements[:2], elements[4:])
+
+    def test_datasets_shuffled(self):
+        cid_datasets = tfds.create_grouped_dataset(self.ds, buffer_size=4)
+        # check that including the buffer shuffles the values in each cid_ds.
+        for cid_ds in cid_datasets:
+            self.assertNotEqual(
+                list(cid_ds.take(20).as_numpy_iterator()),
+                list(cid_ds.take(20).as_numpy_iterator()),
+            )
+
+
+class TestCreateChoicesDataset(tf.test.TestCase):
+    def test_sample_without_replacement(self):
+        # Test that each class appears exactly examples_per_class times
+        num_classes = 5
+        examples_per_class = 2
+        dataset = tfds.create_choices_dataset(num_classes, examples_per_class)
+        elements = list(dataset.take(num_classes * examples_per_class).as_numpy_iterator())
+        unique_elements = set(elements)
+        self.assertLen(unique_elements, num_classes)
```

Comment:
why is examples_per_class irrelevant in this assert? shouldn't the len be num_classes * examples_per_class?